raft.go

```diff
@@ -680,6 +691,94 @@ func (r *Raft) quorumSize() int {
 	return voters/2 + 1
 }
 
+// restoreUserSnapshot is used to manually consume an external snapshot, such
+// as if restoring from a backup. We will use the current Raft configuration,
+// not the one from the snapshot, so that we can restore into a new cluster. We
+// will also use the higher of the index of the snapshot, or the current index,
+// and then add 1 to that, so we force a new state with a hole in the Raft log,
+// so that the snapshot will be sent to followers and used for any new joiners.
+// This can only be run on the leader, and returns a future that can be used to
+// block until complete.
+func (r *Raft) restoreUserSnapshot(future *userRestoreFuture) {
+	defer metrics.MeasureSince([]string{"raft", "restoreUserSnapshot"}, time.Now())
+	err := func() error {
+		// Sanity check the version.
+		version := future.meta.Version
+		if version < SnapshotVersionMin || version > SnapshotVersionMax {
+			return fmt.Errorf("unsupported snapshot version %d", version)
+		}
+
+		// We don't support snapshots while there's a config change
+		// outstanding since the snapshot doesn't have a means to
+		// represent this state.
+		committedIndex := r.configurations.committedIndex
+		latestIndex := r.configurations.latestIndex
+		if committedIndex != latestIndex {
+			return fmt.Errorf("cannot restore snapshot now, wait until the configuration entry at %v has been applied (have applied %v)",
+				latestIndex, committedIndex)
+		}
+
+		// We will overwrite the snapshot metadata with the current term,
+		// an index that's greater than the current index, or the last
+		// index in the snapshot. It's important that we leave a hole in
+		// the index so we know there's nothing in the Raft log there and
+		// replication will fault and send the snapshot.
+		term := r.getCurrentTerm()
+		lastIndex := r.getLastIndex()
+		if future.meta.Index > lastIndex {
+			lastIndex = future.meta.Index
+		}
+		lastIndex++
+
+		// Dump the snapshot. Note that we use the latest configuration,
+		// not the one that came with the snapshot.
+		sink, err := r.snapshots.Create(version, lastIndex, term,
+			r.configurations.latest, r.configurations.latestIndex, r.trans)
+		if err != nil {
+			return fmt.Errorf("failed to create snapshot: %v", err)
+		}
+		n, err := io.Copy(sink, future.reader)
+		if err != nil {
+			sink.Cancel()
+			return fmt.Errorf("failed to write snapshot: %v", err)
+		}
+		if n != future.meta.Size {
+			sink.Cancel()
+			return fmt.Errorf("failed to write snapshot, size didn't match (%d != %d)", n, future.meta.Size)
+		}
+		if err := sink.Close(); err != nil {
+			return fmt.Errorf("failed to close snapshot: %v", err)
+		}
+		r.logger.Printf("[INFO] raft: Copied %d bytes to local snapshot", n)
+
+		// Restore the snapshot into the FSM. If this fails we are in a
+		// bad state so we panic to take ourselves out.
+		fsm := &restoreFuture{ID: sink.ID()}
+		fsm.init()
+		select {
+		case r.fsmRestoreCh <- fsm:
```

Comment:
An interesting problem with the fsmRestoreCh is that we can't promise an ordering with `r.fsmCommitCh`. This could cause issues if we restore and then older commits are applied on top of it. I think we may need to fix the `runFSM` to use a single fixed FIFO queue so that we have reliable ordering.